staging/src/k8s.io/sample-controller/controller_test.go

```diff
@@ -0,0 +1,310 @@
+
+var (
+	alwaysReady        = func() bool { return true }
+	noResyncPeriodFunc = func() time.Duration { return 0 }
+)
+
+type fixture struct {
+	t *testing.T
+
+	client     *fake.Clientset
+	kubeclient *k8sfake.Clientset
+	// Objects to put in the store.
+	fooLister        []*samplecontroller.Foo
+	deploymentLister []*apps.Deployment
+	// Actions expected to happen on the client.
+	kubeactions []core.Action
+	actions     []core.Action
+	// Objects from here preloaded into NewSimpleFake.
+	kubeobjects []runtime.Object
+	objects     []runtime.Object
+}
+
+func newFixture(t *testing.T) *fixture {
+	f := &fixture{}
+	f.t = t
+	f.objects = []runtime.Object{}
+	f.kubeobjects = []runtime.Object{}
+	return f
+}
+
+func newFoo(name string, replicas *int32) *samplecontroller.Foo {
+	return &samplecontroller.Foo{
+		TypeMeta: metav1.TypeMeta{APIVersion: samplecontroller.SchemeGroupVersion.String()},
+		ObjectMeta: metav1.ObjectMeta{
+			UID:       uuid.NewUUID(),
+			Name:      name,
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: samplecontroller.FooSpec{
+			DeploymentName: fmt.Sprintf("%s-deployment", name),
+			Replicas:       replicas,
+		},
+	}
+}
+
+func (f *fixture) newController() (*Controller, informers.SharedInformerFactory, kubeinformers.SharedInformerFactory) {
+	f.client = fake.NewSimpleClientset(f.objects...)
+	f.kubeclient = k8sfake.NewSimpleClientset(f.kubeobjects...)
+
+	i := informers.NewSharedInformerFactory(f.client, noResyncPeriodFunc())
+	k8sI := kubeinformers.NewSharedInformerFactory(f.kubeclient, noResyncPeriodFunc())
+
+	c := NewController(f.kubeclient, f.client,
+		k8sI.Apps().V1beta2().Deployments(), i.Samplecontroller().V1alpha1().Foos())
+
+	c.foosSynced = alwaysReady
+	c.deploymentsSynced = alwaysReady
+	c.recorder = &record.FakeRecorder{}
+
+	for _, f := range f.fooLister {
+		i.Samplecontroller().V1alpha1().Foos().Informer().GetIndexer().Add(f)
+	}
+
+	for _, d := range f.deploymentLister {
+		k8sI.Apps().V1beta2().Deployments().Informer().GetIndexer().Add(d)
+	}
+
+	return c, i, k8sI
+}
+
+func (f *fixture) run(fooName string) {
+	f.runController(fooName, true, false)
+}
+
+func (f *fixture) runExpectError(fooName string) {
+	f.runController(fooName, true, true)
+}
+
+func (f *fixture) runController(fooName string, startInformers bool, expectError bool) {
+	c, i, k8sI := f.newController()
+	if startInformers {
+		stopCh := make(chan struct{})
+		defer close(stopCh)
+		i.Start(stopCh)
+		k8sI.Start(stopCh)
+	}
+
+	err := c.syncHandler(fooName)
+	if !expectError && err != nil {
+		f.t.Errorf("error syncing foo: %v", err)
+	} else if expectError && err == nil {
+		f.t.Error("expected error syncing foo, got nil")
+	}
+
+	actions := filterInformerActions(f.client.Actions())
+	for i, action := range actions {
+		if len(f.actions) < i+1 {
+			f.t.Errorf("%d unexpected actions: %+v", len(actions)-len(f.actions), actions[i:])
+			break
+		}
+
+		expectedAction := f.actions[i]
+		checkAction(expectedAction, action, f.t)
+	}
```

Comment:
Add a note saying we check `f.actions[i]` against `actions[i]`, as f.actions is an **ordered list** of expected actions. Perhaps that note should be on the fixture struct itself 😄

Reply:
I addressed all comments except this one. `fixture` already has comments for its fields.

But since you waited this comment and you familiar with code, I can assume for most people it will be even harder to understand the code.

The problem I can't came with good comment, maybe you can help?

Thanks.